org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/IndexManager.java

```diff
@@ -1654,7 +1652,18 @@ private MetaIndex loadMetaIndexIfNeeded() throws IOException {
 				if (VERBOSE) {
 					Util.verbose("-> load existing meta-index: "+indexLocation+" path: "+INDEX_META_CONTAINER); //$NON-NLS-1$ //$NON-NLS-2$
 				}
-				this.metaIndex = new MetaIndex(new Index(indexLocation, INDEX_META_CONTAINER, true));
+				try {
+					this.metaIndex = new MetaIndex(new Index(indexLocation, INDEX_META_CONTAINER, true));
+				} catch (IOException e) {
+					Util.log(e, "Failed to read saved meta index, re-creating"); //$NON-NLS-1$
+					if (VERBOSE) {
+						Util.verbose("-> failed to read saved meta-index: "+indexLocation+" path: "+INDEX_META_CONTAINER); //$NON-NLS-1$ //$NON-NLS-2$
+						Util.verbose("-> re-create meta-index: "+indexLocation+" path: "+INDEX_META_CONTAINER); //$NON-NLS-1$ //$NON-NLS-2$
+					}
+					indexLocation.delete();
+					this.metaIndex = new MetaIndex(new Index(indexLocation, INDEX_META_CONTAINER, false));
+					updateIndexState(indexLocation, REUSE_STATE);
```

Comment:
@gayanper : I'm not sure about this `REUSE_STATE`, I just copied code from the branch above where the new index is created.

Reply:
We used this this state to avoid normal index handling code not to delete this index and try to rebuild it.

Reply:
OK, so if index is broken (or incompatible) it is OK to delete it, so REUSE_STATE would trigger rebuild after delete, right?

Reply:
If we are doing a full re-index, then that should be fine.